validator/src/domain/event.rs

```diff
@@ -0,0 +1,244 @@
+use std::collections::HashMap;
+
+use chrono::{DateTime, Utc};
+use num_traits::CheckedSub;
+use serde::{Deserialize, Serialize};
+
+use domain::balances_map::get_balances_after_fees_tree;
+use domain::validator::message::Accounting;
+use domain::{AdUnit, BalancesMap, BigNum, Channel, ChannelId, DomainError};
+
+#[serde(tag = "type", rename_all = "SCREAMING_SNAKE_CASE")]
+#[derive(Serialize, Deserialize)]
+pub enum Event {
+    #[serde(rename_all = "camelCase")]
+    Impression {
+        publisher: String,
+        // clippy warning for big size difference, because of field
+        ad_unit: Box<AdUnit>,
+    },
+    ImpressionWithCommission {
+        earners: Vec<Earner>,
+    },
+    /// only the creator can send this event
+    UpdateImpressionPrice {
+        price: BigNum,
+    },
+    /// only the creator can send this event
+    Pay {
+        outputs: HashMap<String, BigNum>,
+    },
+    /// only the creator can send this event
+    PauseChannel,
+    /// only the creator can send this event
+    Close,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Earner {
+    #[serde(rename = "publisher")]
+    pub address: String,
+    pub promilles: u64,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct EventAggregate {
+    pub channel_id: ChannelId,
+    pub created: DateTime<Utc>,
+    pub events: HashMap<String, AggregateEvents>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AggregateEvents {
+    pub event_counts: HashMap<String, BigNum>,
+    pub event_payouts: HashMap<String, BigNum>,
+}
+
+#[allow(dead_code)]
+fn merge_aggrs(
+    accounting: &Accounting,
+    aggregates: &[EventAggregate],
+    channel: &Channel,
+) -> Result<(BalancesMap, Accounting), DomainError> {
+    let deposit = channel.deposit_amount.clone();
+
+    let last_event_aggregate = [accounting.last_event_aggregate]
+        .iter()
+        .chain(aggregates.iter().map(|aggr| &aggr.created))
+        .max_by(|lhs, rhs| lhs.cmp(rhs))
```

Comment:
why not `max`?

Reply:
You are right, overlooked this. I think it's part of the refactor later on, because I switched it to mapping the datetimes instead of using the aggr.